File: single_linkage.py
#!/usr/bin/python
"""
Single linkage clustering
Usage:
    python single_linkage.py --input mx.txt --output clusters.txt --cutoff 0.2

Author - Felipe González Casabianca
2024, April 03
"""


import argparse


def get_single_linkage_clusters(input_path, cut_off):
    """
    Method that computes the single linkage clusters from the given sparse matrix file.
    Note: Cutoff is assumed to be inclusive, that is: clusters will not be merged if the distance
    is larger or equal to the given cutoff
    """

    clusters_id = {}
    cluster_members = {}
    labels_set = set()

    num_clusters = 0
    with open(input_path, "r") as file:
        for line in file:
            label1, label2, distance = line.strip().split()

            # Stop condition
            if len(labels_set) > 0 and len(labels_set) == len(clusters_id):
                # Done
                print("Finished")
                break

            if label1 == label2:
                labels_set.add(label1)
            else:

                distance = float(distance)

                # Checks for distance and cutoff
                if distance >= cut_off:
                    continue

                # Checks if labels already have been assigned
                in1 = label1 in clusters_id
                in2 = label2 in clusters_id

                # Cases
                if in1 and not in2:
                    clusters_id[label2] = clusters_id[label1]
                    cluster_members[clusters_id[label1]].add(label2)
                elif in2 and not in1:
                    clusters_id[label1] = clusters_id[label2]
                    cluster_members[clusters_id[label2]].add(label1)
                elif not in2 and not in1:
                    clusters_id[label1] = num_clusters
                    clusters_id[label2] = num_clusters
                    cluster_members[num_clusters] = set([label1, label2])
                    num_clusters += 1
                else:
                    if clusters_id[label1] == clusters_id[label2]:
                        continue
                    else:
                        # merges
                        to_be_merged = clusters_id[label1]
                        to_be_deleted = clusters_id[label2]
                        cluster_members[to_be_merged].update(
                            cluster_members[to_be_deleted]
                        )
                        # Updates
                        for k in cluster_members[to_be_deleted]:
                            clusters_id[k] = to_be_merged

                        del cluster_members[to_be_deleted]

    # Adds isolated labels
    for lab in labels_set.difference(clusters_id.keys()):
        clusters_id[lab] = num_clusters
        cluster_members[num_clusters] = set([lab])
        num_clusters += 1

    labels_list = list(labels_set)

    # Cleans so cluster labels are continuos
    new_cluster_ids = {}
    new_num_clusters = 0
    for _, value in cluster_members.items():
        new_cluster_ids.update({key: new_num_clusters for key in value})
        new_num_clusters += 1

    cluster_labels = [new_cluster_ids[k] for k in labels_list]

    return labels_list, cluster_labels


def export_clusters(output_path, labels_list, cluster_labels):
    sorted_labels = sorted(zip(cluster_labels, labels_list), key=lambda x: (x[0], x[1]))
    with open(output_path, "w") as file:
        for cluster_label, label in sorted_labels:
            file.write(f"{cluster_label}\t{label}\n")


def main():
    parser = argparse.ArgumentParser(
        description="Compute single linkage clusters for the output of `usearch -calc_distmx`."
    )
    parser.add_argument(
        "--input", required=True, help="Path to the input sparse matrix file"
    )
    parser.add_argument("--output", required=True, help="Path to the output file")
    parser.add_argument(
        "--cutoff", type=float, required=True, help="Distance cutoff for clustering"
    )
    args = parser.parse_args()
    labels_list, cluster_labels = get_single_linkage_clusters(args.input, args.cutoff)
    export_clusters(args.output, labels_list, cluster_labels)


if __name__ == "__main__":
    main()

File: single_linkage.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

// clusterInfo holds information about a single cluster member
type clusterInfo struct {
	Label     string
	ClusterID int
}

// getSingleLinkageClusters reads pairwise distances from the input file,
// forms clusters based on the cutoff distance,
// and returns cluster members and their IDs
func getSingleLinkageClusters(inputPath string, cutOff float64, includeEqual bool) ([]clusterInfo, error) {
	clustersID := make(map[string]int)
	clusterMembers := make(map[int]map[string]bool)
	labelsSet := make(map[string]bool)

	numClusters := 0
	file, err := os.Open(inputPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) < 3 {
			continue // Skip lines that don't have enough parts
		}
		label1, label2, distanceStr := parts[0], parts[1], parts[2]

		labelsSet[label1] = true
		labelsSet[label2] = true

		distance, err := strconv.ParseFloat(distanceStr, 64)
		if err != nil {
			return nil, err
		}

		// Comparison with the cutoff depends on the `--includeequal` flag
		if includeEqual && distance > cutOff || !includeEqual && distance >= cutOff {
			continue
		}

		in1, ok1 := clustersID[label1]
		in2, ok2 := clustersID[label2]

		if !ok1 && !ok2 {
			clusterID := numClusters
			clustersID[label1] = clusterID
			clustersID[label2] = clusterID
			clusterMembers[clusterID] = map[string]bool{label1: true, label2: true}
			numClusters++
		} else if ok1 && !ok2 {
			clustersID[label2] = in1
			clusterMembers[in1][label2] = true
		} else if !ok1 && ok2 {
			clustersID[label1] = in2
			clusterMembers[in2][label1] = true
		} else if in1 != in2 {
			for label := range clusterMembers[in2] {
				clustersID[label] = in1
				clusterMembers[in1][label] = true
			}
			delete(clusterMembers, in2)
		}
	}

	// Reassign cluster IDs to be zero-based and sequential
	sequentialClusters := reassignClusterIDs(clustersID)

	return sequentialClusters, nil
}

// Complete linkage clustering
func getCompleteLinkageClusters(inputPath string, cutOff float64, includeEqual bool) ([]clusterInfo, error) {
	clustersID := make(map[string]int)
	clusterMembers := make(map[int]map[string]bool)
	maxDistances := make(map[int]float64) // track maximum distances per cluster
	// labelsSet := make(map[string]bool)

	numClusters := 0
	file, err := os.Open(inputPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) < 3 {
			continue // Skip insufficient data
		}
		label1, label2, distanceStr := parts[0], parts[1], parts[2]

		distance, err := strconv.ParseFloat(distanceStr, 64)
		if err != nil {
			return nil, err
		}

		in1, ok1 := clustersID[label1]
		in2, ok2 := clustersID[label2]

		if !ok1 && !ok2 {
			// Both labels are new, create a new cluster
			clusterID := numClusters
			clustersID[label1] = clusterID
			clustersID[label2] = clusterID
			clusterMembers[clusterID] = map[string]bool{label1: true, label2: true}
			maxDistances[clusterID] = distance
			numClusters++
		} else if ok1 && !ok2 {
			clustersID[label2] = in1
			clusterMembers[in1][label2] = true
			updateMaxDistance(maxDistances, distance, in1)
		} else if !ok1 && ok2 {
			clustersID[label1] = in2
			clusterMembers[in2][label1] = true
			updateMaxDistance(maxDistances, distance, in2)
		} else if in1 != in2 {
			// Based on maxDistances, decide whether to merge or not
			shouldMerge := (includeEqual && maxDistances[in1] <= cutOff) || (!includeEqual && maxDistances[in1] < cutOff)
			if shouldMerge {
				// Merge the clusters
				mergeClusters(clusterMembers, clustersID, in1, in2)
				// After merging, merge their maxDistances and recompute
				maxDistances[in1] = max(maxDistances[in1], maxDistances[in2])
				delete(maxDistances, in2)
			}
		}
	}

	// Reassign cluster IDs to be zero-based and sequential
	sequentialClusters := reassignClusterIDs(clustersID)

	return sequentialClusters, nil
}


// Function to reassign cluster IDs sequentially
func reassignClusterIDs(clustersID map[string]int) []clusterInfo {
	newID := 0
	oldToNewID := make(map[int]int)
	var clusters []clusterInfo

	for label, oldID := range clustersID {
		if _, exists := oldToNewID[oldID]; !exists {
			oldToNewID[oldID] = newID
			newID++
		}
		clusters = append(clusters, clusterInfo{Label: label, ClusterID: oldToNewID[oldID]})
	}

	return clusters
}

// exportClusters writes the cluster members and their IDs to the output file, sorted first by cluster ID and then by label
func exportClusters(outputPath string, clusters []clusterInfo) error {
	file, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer file.Close()

	// Sort clusters by ClusterID, then by Label
	sort.Slice(clusters, func(i, j int) bool {
		if clusters[i].ClusterID == clusters[j].ClusterID {
			return clusters[i].Label < clusters[j].Label
		}
		return clusters[i].ClusterID < clusters[j].ClusterID
	})

	for _, cluster := range clusters {
		if _, err := fmt.Fprintf(file, "%d\t%s\n", cluster.ClusterID, cluster.Label); err != nil {
			return err
		}
	}

	return nil
}

func main() {
	input := flag.String("input", "", "Path to the input file containing pairwise distances")
	output := flag.String("output", "", "Path to the output file for cluster assignments")
	cutoff := flag.Float64("cutoff", 0.0, "Distance cutoff for clustering (must be greater than 0)")
	includeEqual := flag.Bool("includeequal", true, "Include distances equal to cutoff in clustering (default is true; set it to false for strictly greater than cutoff)")

	flag.Parse()

	if *input == "" || *output == "" || *cutoff == 0.0 {
		log.Println("Input, output, and cutoff parameters are required.")
		flag.Usage()
		return
	}

	clusters, err := getSingleLinkageClusters(*input, *cutoff, *includeEqual)
	if err != nil {
		log.Fatalf("Error processing clusters: %v", err)
	}

	if err := exportClusters(*output, clusters); err != nil {
		log.Fatalf("Error exporting clusters: %v", err)
	}
}
